### PR TITLE
Java interview Vlad Smarttech

### PR DESCRIPTION
**Pull request by hedinant**

### CODE REVIEW

File: TaskManager.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
+@Component
+public class TaskManager {
+   private static TaskManager instance;
+   private LinkedList<Task> queue = new LinkedList<>();
```
Comment (uladzimirkuzniatsov) [inline TaskManager.java]: // initialize by getting it from outside (via constructor)

File: Task.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
+public abstract class Task {
```
Comment (hedinant) [inline Task.java:1]: test

File: TaskHistoryRecord.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
+@Entity
+@Table(name = "taskhistory")
+public class TaskHistoryRecord {
+    @Id
+    private long id;
+    private String record;
+
```
Comment (uladzimirkuzniatsov) [inline TaskHistoryRecord.java:7]: // default constructor

File: TaskManager.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
+@Component
+public class TaskManager {
+    @PersistenceContext
+    private EntityManager entityManager;
+    private static TaskManager instance;
+    private LinkedList<Task> queue = new LinkedList<>();
```
Comment (uladzimirkuzniatsov) [inline TaskManager.java:6]: // init it via constructor?

File: TaskManager.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    private HashSet<Task> running = new HashSet<>();
+    private HashSet<Task> blocked = new HashSet<>();
+    private Thread checker = new Thread(() -> {
+        while (true) {
```
Comment (uladzimirkuzniatsov) [inline TaskManager.java:10]: // how to exit the infinity loop

File: TaskManager.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+            checkRun();
+        }
+    });
+
+    public TaskManager() {
+        instance = this;
+        checker.run();
```
Comment (uladzimirkuzniatsov) [inline TaskManager.java:17]: // .run ? perhaps we need to start the thread by .start?

File: TaskManager.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    public static TaskManager getInstance() {
+        return instance;
+    }
+
+    @Transactional
+    public void runTask(Task task) {
+        if (Collections.disjoint(task.getDependsOn(), running)) {
+            running.add(task);
+            task.setStatus(TaskStatus.Working);
+            new Thread(() -> task.runTask(entityManager)).run();
```
Comment (uladzimirkuzniatsov) [inline TaskManager.java:29]: // be sure entityManager is thread safe

Comment (uladzimirkuzniatsov) [inline TaskManager.java:29]: // .run ? not .start?

File: TaskManager.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+                return;
+            }
+    }
+
+    public boolean isRunning(String name){
+        return running.stream().anyMatch(task -> task.getName() == name);
+    }
+
+    private void checkRun() {
+        while (queue.get(0).getSchedule() <= System.currentTimeMillis()) {
```
Comment (uladzimirkuzniatsov) [inline TaskManager.java:67]: //.get(0) - NPE?

File: TaskManager.java [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+            runTask(queue.removeFirst());
```
Comment (uladzimirkuzniatsov) [inline TaskManager.java:68]: // called transactional method - transaction does not work

File: TaskManager.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+    public void skeduleTask(Task task) {
+        for (int i = 0; i < queue.size(); i++)
+            if (task.getSchedule() < queue.get(i).getSchedule()) {
+                queue.add(i, task);
+                return;
+            }
+    }
+
+    public boolean isRunning(String name){
+        return running.stream().anyMatch(task -> task.getName() == name);
```
Comment (uladzimirkuzniatsov) [inline TaskManager.java:63]: // string compare via .equals

File: TaskManager.java [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+            if (Collections.disjoint(task1.getDependsOn(), running)) {
+                blocked.remove(task1);
+                runTask(task1);
+            }
+        });
+    }
+
+    public void skeduleTask(Task task) {
+        for (int i = 0; i < queue.size(); i++)
+            if (task.getSchedule() < queue.get(i).getSchedule()) {
```
Comment (uladzimirkuzniatsov) [inline TaskManager.java:56]: // .get(i) - NPE, ArrayIndexOutOfBound

File: TaskManager.java (hunk 7 above)
Comment (uladzimirkuzniatsov) [inline TaskManager.java:29]: // manual Thread managing. Better use ExecutorService or at least a Thread Pool

**Pull request closed without merging**